Mathlib/Topology/UniformSpace/MultipliableTendstoUniformly.lean

```diff
@@ -0,0 +1,121 @@
+
+/-!
+# Uniform limits of products of functions
+
+We gather some results about the uniform convergence of the products, in particular those of the
+form `1 + f n x` for a sequence `f n x` of complex numbers.
+
+-/
+
+open Filter Function Complex
+
+variable {α β ι : Type*}
+
+lemma TendstoUniformlyOn.eventually_forall_lt [UniformSpace β] [AddGroup β]
+    [UniformAddGroup β] [LinearOrder β] [OrderTopology β] [AddLeftStrictMono β] [AddRightMono β]
+    {f : ι → α → β} {p : Filter ι} {g : α → β} {K : Set α} {T V : β} (htv : T < V)
+    (hf : TendstoUniformlyOn f g p K) (hg : ∀ x, x ∈ K → g x ≤ T) :
+    ∀ᶠ (n : ι) in p, ∀ x, x ∈ K → f n x < V := by
+  rw [tendstoUniformlyOn_iff_tendsto] at hf
+  simp only [uniformity_eq_comap_neg_add_nhds_zero, tendsto_iff_eventually, eventually_comap,
+    Prod.forall] at *
+  conv at hf => enter [2]; rw [eventually_iff_exists_mem]
+  have hf2 := hf (fun x : β × β => -x.1 + x.2 < -T + V) ⟨{x : β | x < -T + V},
+      IsOpen.mem_nhds (isOpen_gt' (-T + V)) (by simp [htv]),
+      fun y hy a b hab => by simpa [hab] using hy⟩
+  rw [eventually_prod_principal_iff, eventually_iff_exists_mem] at *
+  obtain ⟨v, hv, H⟩ := hf2
+  refine ⟨v, hv, fun y hy x hx => by simpa using (add_lt_add_of_le_of_lt (hg x hx) (H y hy x hx))⟩
+
+lemma TendstoUniformlyOn.eventually_forall_le [UniformSpace β] [AddGroup β]
+    [UniformAddGroup β] [LinearOrder β] [OrderTopology β] [AddLeftStrictMono β] [AddRightMono β]
+    (f : ι → α → β)  {p : Filter ι} {g : α → β} {K : Set α} {T V : β} (htv : T < V)
+    (hf : TendstoUniformlyOn f g p K) (hg : ∀ x, x ∈ K → g x ≤ T) :
+    ∀ᶠ (n : ι) in p, ∀ x, x ∈ K → f n x ≤ V := by
+    filter_upwards [hf.eventually_forall_lt htv hg] with i hi x hx using (hi x hx).le
+
+lemma tendstoUniformlyOn_comp_cexp {p : Filter ι} {f : ι → α → ℂ} {g : α → ℂ}
+    {K : Set α} (hf : TendstoUniformlyOn f g p K) (hg : BddAbove ((fun x => (g x).re) '' K)) :
+    TendstoUniformlyOn (fun n => fun x => cexp (f n x)) (cexp ∘ g) p K := by
+  obtain ⟨T, hT⟩ := hg
+  simp only [mem_upperBounds, Set.mem_image, forall_exists_index, and_imp,
+    forall_apply_eq_imp_iff₂] at hT
+  have h2 := TendstoUniformlyOn.eventually_forall_le (fun n x => (f n x).re) (lt_add_one T)
+    hf.re hT
+  have w2 := tendstoUniformlyOn_univ.mpr <| UniformContinuousOn.comp_tendstoUniformly_eventually
+    {x : ℂ | x.re ≤ T + 1} (fun a => K.restrict (f (a))) (fun b => g b) (by simpa using h2) ?_
+      (UniformContinuousOn.cexp (T + 1)) ((tendstouniformlyOn_iff_restrict).mp hf)
+  · rw [tendstouniformlyOn_iff_restrict, ← tendstoUniformlyOn_univ]
+    exact w2
+  · simp only [Set.mem_setOf_eq, Subtype.forall]
+    exact fun a ha => le_trans (hT a ha) (by aesop)
+
+lemma tendstoUniformlyOn_tprod_of_clog {f : ι → α → ℂ} {K : Set α}
+    (h : ∀ x, x ∈ K → Summable fun n => log (f n x))
+    (hf : TendstoUniformlyOn (fun n a => ∑ i ∈ n, log (f i a))
+    (fun a : α => ∑' n : ι, log (f n a)) atTop K) (hfn : ∀ x, x ∈ K → ∀ n : ι, f n x ≠ 0)
```

Comment:
... and the same for explicit type-ascriptions (`a` can't be anything else since it's the second argument to `f`).